File: frontend/lib/api/status.ts
import { getStoredToken } from "./auth";
import { UserStatus, StatusSlide } from "@/types/status";

const API_URL = process.env.NEXT_PUBLIC_API_URL || "http://localhost:8000";

export interface CreateStatusPayload {
  type: "image" | "text";
  content: string;
  caption?: string;
  background_color?: string;
  font_style?: "modern" | "serif" | "mono" | "bold";
}

export interface StatusFeedResponse {
  items: UserStatus[];
  total: number;
}

/**
 * Posts a new 24h status slide (photo or text).
 */
export async function createStatusSlide(
  payload: CreateStatusPayload,
  token?: string
): Promise<StatusSlide> {
  const authToken = token || getStoredToken();
  const res = await fetch(`${API_URL}/api/v1/status`, {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
      Authorization: `Bearer ${authToken}`,
    },
    body: JSON.stringify(payload),
  });

  if (!res.ok) {
    const err = await res.json().catch(() => ({}));
    throw new Error(err.detail || "Failed to post status.");
  }

  const data = await res.json();
  return {
    id: data.id,
    type: data.type,
    content: data.content,
    caption: data.caption,
    backgroundColor: data.background_color,
    fontStyle: data.font_style,
    createdAt: data.created_at,
  };
}

/**
 * Fetches the active status feed visible to the user (the user + connected contacts).
 */
export async function getActiveStatuses(token?: string): Promise<UserStatus[]> {
  const authToken = token || getStoredToken();
  if (!authToken) return [];

  const res = await fetch(`${API_URL}/api/v1/status`, {
    headers: {
      Authorization: `Bearer ${authToken}`,
    },
  });

  if (!res.ok) {
    return [];
  }

  const data = await res.json();
  const items = data.items || [];

  // Map to frontend UserStatus model
  return items.map((u: any) => ({
    id: u.id,
    userId: u.user_id,
    userName: u.user_name,
    userAvatar: u.user_avatar,
    userInitials: u.user_initials,
    isMe: u.is_me,
    viewed: u.viewed,
    lastUpdated: u.last_updated,
    slides: (u.slides || []).map((s: any) => ({
      id: s.id,
      type: s.type,
      content: s.content,
      caption: s.caption,
      backgroundColor: s.background_color,
      fontStyle: s.font_style,
      createdAt: s.created_at,
    })),
  }));
}

/**
 * Deletes an individual status slide.
 */
export async function deleteStatusSlide(slideId: string, token?: string): Promise<void> {
  const authToken = token || getStoredToken();
  const res = await fetch(`${API_URL}/api/v1/status/${slideId}`, {
    method: "DELETE",
    headers: {
      Authorization: `Bearer ${authToken}`,
    },
  });

  if (!res.ok) {
    const err = await res.json().catch(() => ({}));
    throw new Error(err.detail || "Failed to delete slide.");
  }
}

/**
 * Deletes all status slides belonging to the authenticated user.
 */
export async function deleteMyStatus(token?: string): Promise<void> {
  const authToken = token || getStoredToken();
  const res = await fetch(`${API_URL}/api/v1/status`, {
    method: "DELETE",
    headers: {
      Authorization: `Bearer ${authToken}`,
    },
  });

  if (!res.ok) {
    const err = await res.json().catch(() => ({}));
    throw new Error(err.detail || "Failed to delete status.");
  }
}

File: backend/services/user-service/app/schemas/status.py
"""
Pydantic schemas for 24-hour Status / Stories feature.
"""
from datetime import datetime
from typing import List, Optional
from pydantic import BaseModel, Field


class CreateStatusPayload(BaseModel):
    """Payload to post a new status slide (photo or text)."""
    type: str = Field("text", description="'image' | 'text'")
    content: str = Field(..., min_length=1, max_length=5000, description="Cloudinary media URL or text content")
    caption: Optional[str] = Field(None, max_length=500, description="Optional photo caption")
    background_color: Optional[str] = Field(None, description="Tailwind gradient or color code")
    font_style: Optional[str] = Field("modern", description="'modern' | 'serif' | 'mono' | 'bold'")


class StatusSlideResponse(BaseModel):
    """Individual slide in a user's 24h status story."""
    id: str
    type: str
    content: str
    caption: Optional[str] = None
    background_color: Optional[str] = None
    font_style: Optional[str] = None
    created_at: datetime
    expires_at: datetime


class UserStatusResponse(BaseModel):
    """Aggregated status story for a single user."""
    id: str
    user_id: str
    user_name: str
    user_avatar: Optional[str] = None
    user_initials: str
    is_me: bool = False
    slides: List[StatusSlideResponse]
    viewed: bool = False
    last_updated: datetime


class StatusFeedResponse(BaseModel):
    """List of active statuses visible to the user."""
    items: List[UserStatusResponse]
    total: int


class ActionSuccessResponse(BaseModel):
    """Generic status action response."""
    status: str = "ok"
    message: str
